### PR TITLE
Custom error messages for IO with nonexistent files

### PR DESCRIPTION
**Pull request by vuule**
## Description
Closes https://github.com/rapidsai/cudf/issues/12311; closes https://github.com/rapidsai/cudf/issues/9564
We return a somewhat cryptic error when opening a file that does not exist: "Cannot query file size".

With this change, we report whether the file exists, or, if the file does exist, what the errno value is after `open`.
Also added a check for the output files' directory in `file_sink`. This check is now also included in `file_wrapper`, just in case initialization order changes at some point.
Now we should always correctly report missing output file directory and missing input files.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/io/utilities/data_sink.cpp [hunk 1]
```diff
@@ -34,7 +36,13 @@ class file_sink : public data_sink {
...
   {
     _output_stream.open(filepath, std::ios::out | std::ios::binary | std::ios::trunc);
-    CUDF_EXPECTS(_output_stream.is_open(), "Cannot open output file");
+    if (!_output_stream.is_open()) {
+      auto const dir_path = std::filesystem::path(filepath).parent_path();
+      if (not std::filesystem::exists(dir_path)) {
+        CUDF_FAIL("Cannot open output file; directory does not exist");
+      }
+      CUDF_FAIL("Cannot open output file");
+    }
```
Comment (karthikeyann) [inline cpp/src/io/utilities/data_sink.cpp:45]: would it help to make this code also similar to `cpp/src/io/utilities/file_io_utilities.cpp` ? That seems more detailed.

Comment (vuule) [reply cpp/src/io/utilities/data_sink.cpp:48]: added errno. I originally excluded it because I didn't know if `ofstream::open` sets it.

Comment (ttnghia) [reply cpp/src/io/utilities/data_sink.cpp:48]: Should we move this into `file_io_utilities`? As it should contain all file IO.

Comment (mythrocks) [reply cpp/src/io/utilities/data_sink.cpp:48]: TIL `std::strerror`

I had to look up where the allocation happens for the string returned by `std::strerror()`. I'm not convinced it is threadsafe, given that it's allowed to reuse the same allocated space to return strings across threads.

Should we consider switching this to `std::strerror_s()`?

File: cpp/src/io/utilities/file_io_utilities.cpp [hunk 2]
```diff
@@ -21,30 +21,43 @@
...
-  struct stat st;
-  CUDF_EXPECTS(fstat(file_descriptor, &st) != -1, "Cannot query file size");
-  return static_cast<size_t>(st.st_size);
+  auto const fd = open(filepath.c_str(), flags, mode);
+  if (fd != -1) { return fd; }
+
+  // save errno because it may be overwritten by subsequent calls
+  auto const err = errno;
+
+  if (auto const path = std::filesystem::path(filepath); flags & O_CREAT) {
```
Comment (mythrocks) [inline cpp/src/io/utilities/file_io_utilities.cpp:40]: Noice.

**Review by mythrocks — COMMENTED** [review]
LGTM, save for the potential thread-safety issue.

File: cpp/src/io/utilities/data_sink.cpp (hunk 1 above)
Comment (vuule) [reply cpp/src/io/utilities/data_sink.cpp:48]: great catch! I'll look into `strerror_s`.

Comment (vuule) [reply cpp/src/io/utilities/data_sink.cpp:48]: > Should we move this into file_io_utilities?

I don't think `file_sink` belongs in utilities, but I'll try to reuse this error checking and place the common code in `file_io_utilities`.

Comment (vuule) [reply cpp/src/io/utilities/data_sink.cpp:48]: I should have just put a mutex in here. `strerror_s` portability is horrendous.
Our compiler does not support it, so I have to use `strerror_r`. Aaand, this one is problematic as well, see https://www.club.cc.cmu.edu/~cmccabe/blog_strerror.html

**Review by ttnghia — APPROVED** [review]

**Review by karthikeyann — APPROVED** [review]
TIL `[[noreturn]]`

**Pull request merged**